pkg/allocator/cache.go

```diff
@@ -200,6 +200,19 @@ func (c *cache) stop() {
 	c.stopWatchWg.Wait()
 }
 
+// drain emits a deletion event for all known IDs. It must be called after the
+// cache has been stopped, to ensure that no new events can be received afterwards.
+func (c *cache) drain() {
```

Comment:
Before starting to drain the identities we wait for all incoming events to have been processed and the watch stream to be interrupted (through the waitgroup), in order to prevent that any new events are received afterwards. And all reconnection operations are performed through a per-cluster controller, hence we are guaranteed they are always executed by the same goroutine.